### PR TITLE
add long-term memory and support toolllama as brains;

### PR DESCRIPTION
**Pull request by XuHwang**
- [x] Add long-term memory
- [x]  Support partial matching of tool names
- [x] Verified that the added system prompt does not affect plan-first agent
- [x] Add Chat-Rec in user simulator

### CODE REVIEW

**Comment by XuHwang** [comment]
@microsoft-github-policy-service agree

**Comment by XuHwang** [comment]
@Leavingseason  Looking forward to your careful review~

File: llm4crs/agent_plan_first_openai.py [hunk 1]
```diff
@@ -52,22 +60,40 @@ def _check_plan_legacy(self, plans: Dict) -> bool:
                 return False
         return True
 
-    def run(self, inputs: str):
+    def run(self, inputs: str, user_profile: UserProfileMemory=None):
```
Comment (Leavingseason) [inline llm4crs/agent_plan_first_openai.py:63]: Have you tested whether the original performance (no life-long experience, i.e., when long-term memory is disable and only short-term memory exists) remains good after adding the user profile mechanism?

File: eval/user_simulator.py [hunk 2]
```diff
@@ -140,129 +144,239 @@ def __init__(
...
+        reply = self.engine.call(
+            user_prompt=usr_prompt,
+            sys_prompt=sys_prompt,
+            max_tokens=256,
+            temperature=0.8
+        )
+        return reply
+
+
+class ChatRec:
```
Comment (Leavingseason) [inline eval/user_simulator.py:175]: add a reference to ChatRec

Comment (XuHwang) [reply eval/user_simulator.py:175]: Added now

File: llm4crs/utils/sql.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import sqlparse
+
+def extract_columns_from_where(sql_query):
```
[the remark this answers is not in the document]
Comment (XuHwang) [reply llm4crs/utils/sql.py:6]: Okay~

File: llm4crs/agent_plan_first_openai.py (hunk 1 above)
Comment (XuHwang) [reply llm4crs/agent_plan_first_openai.py:63]: Yes, the introduction of user profile does not affects original performance.

File: llm4crs/memory/memory.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+_FEW_SHOT_EXAMPLES = \
+"""
+> Conversations
+User: My history is ITEM-1, ITEM-2, ITEM-3. Now I want something new.
+Assistent: Based on your preference, I recommend you ITEM-17, ITEM-19, ITEM-30.
+User: I don't like those items, give me more options.
+Assistent: Based on your feedbacks, I recommend you ITEM-5, ITEM-100.
+User: I think ITEM-100 may be very interesting. I may like it. 
+> Profiles
+{"history": ["ITEM-1", "ITEM-2", "ITEM-3"], "like": ["ITEM-100"], "unwanted": ["ITEM-17", "ITEM-19", "ITEM-30"]}
```
[the remark this answers is not in the document]
Comment (XuHwang) [reply llm4crs/memory/memory.py:15]: It consists of the inputs to the recommendation tool, including the user's long-term preferences.

**Review by Leavingseason — APPROVED** [review]

**Pull request merged**